types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,658 @@
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <wlr/backend/drm.h>
+#include <wlr/backend/multi.h>
+#include <wlr/types/wlr_drm_lease_v1.h>
+#include <wlr/util/log.h>
+#include "backend/drm/drm.h"
+#include "drm-lease-v1-protocol.h"
+#include "util/signal.h"
+#include "util/global.h"
+
+#define DRM_LEASE_DEVICE_V1_VERSION 1
+
+static struct wp_drm_lease_device_v1_interface lease_device_impl;
+static struct wp_drm_lease_connector_v1_interface lease_connector_impl;
+static struct wp_drm_lease_request_v1_interface lease_request_impl;
+static struct wp_drm_lease_v1_interface lease_impl;
+
+static struct wlr_drm_lease_device_v1 *drm_lease_device_v1_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+				&wp_drm_lease_device_v1_interface, &lease_device_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static struct wlr_drm_lease_connector_v1 *
+drm_lease_connector_v1_from_resource(struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+				&wp_drm_lease_connector_v1_interface, &lease_connector_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static struct wlr_drm_lease_request_v1 *drm_lease_request_v1_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+				&wp_drm_lease_request_v1_interface, &lease_request_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static struct wlr_drm_lease_v1 *drm_lease_v1_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+				&wp_drm_lease_v1_interface, &lease_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void drm_lease_v1_destroy(struct wlr_drm_lease_v1 *lease) {
+	if (!lease) {
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Destroying lease %"PRIu32, lease->lessee_id);
+
+	wlr_drm_lease_v1_revoke(lease);
```

Comment:
We should probably only do this if `lease->lessee_id != 0`. Or do the check inside `wlr_drm_lease_v1_revoke`, as you prefer.